content/blog/buildkite-cd.md

```diff
@@ -0,0 +1,136 @@
+---
+author: "Rob Ashton"
+date: 2022-09-27
+title: Continuous Delivery with Buildkite
+tags:
+ - Buildkite
+ - CD
+ - deployment
+ - DevOps
+---
+
+# Introduction
+
+Our deployments at RESIDE typically consist of a python script using [constellation](https://github.com/reside-ic/constellation) to bring up a set of docker containers each of which contains a separate part of the application. We might have one docker image for the web app, one for the database, one for redis, one for an R backend API, and a worker image. We use [Buildkite](https://buildkite.com/) as part of our CI process to build these images and test them.
+
+When we want to deploy a new feature to production our process typically involves:
+1. ssh onto the staging server
+1. Run deployment script
+1. Manually test the app for any issues
+1. ssh onto the production server
+1. Deploy to production
+1. Manually test the app for any issues
+
+This creates barriers to deployments, there are tedious manual steps involved and if you do not deploy regularly it is not always obvious what server the app is running on, what script needs to be run, and how it is run. Additionally, we want to limit access to the machines to avoid accidental breakages. We have researchers on the team who don't work with the command line regularly and want to be able to deploy changes to their part of the code to see the effect it will have on the full app.
+
+We wanted a way to
+* Reduce the barrier to deployment making it easy and quick to encourage frequent integration and deployment
+* Deploy automatically to a staging server we can always review the state of the current master/main
+* Allow science to redeploy and see their code changes without having to go through the development team
+
+This blog post covers how we have used Buildkite within RESIDE to support continuous delivery of new features to staging environments. This does not cover details of the deployment script itself but how to set up Buildkite agents and configure a pipeline so that you can deploy via running a build, trigger it automatically and deploy specific tags of your docker images through environment variables.
+
+# Setting up Buildkite agent
+
+We want to create a separate agent from our normal build agents so that a deployment does not have to wait on any long-running builds. Buildkite agents can be configured to listen to specific queues. The default queue is called `default`, jobs added without a specified queue go onto the `default` queue and Buildkite agents without a specified queue pull jobs from this queue. We can create an agent which listens to a different queue by setting the `queue` tag in the [agent configuration](https://buildkite.com/docs/agent/v3/configuration). For example set `tags="queue=deploy"` to pull jobs from the `deploy` queue.
```

Comment:
```suggestion
We want to create a separate agent from our normal build agents so that a deployment does not have to wait on any long-running builds. Buildkite agents can be configured to listen to specific queues. The default queue is called `default` - jobs added without a specified queue go onto the `default` queue and Buildkite agents without a specified queue pull jobs from this queue. We can create an agent which listens to a different queue by setting the `queue` tag in the [agent configuration](https://buildkite.com/docs/agent/v3/configuration). For example set `tags="queue=deploy"` to pull jobs from the `deploy` queue.
```